sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/MessageFlux.java

```diff
@@ -0,0 +1,971 @@
+import reactor.core.CoreSubscriber;
+import reactor.core.Disposable;
+import reactor.core.Disposables;
+import reactor.core.Exceptions;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.FluxOperator;
+import reactor.core.publisher.Mono;
+import reactor.core.publisher.Operators;
+import reactor.core.scheduler.Schedulers;
+import reactor.util.concurrent.Queues;
+
+import java.time.Duration;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Queue;
+import java.util.concurrent.RejectedExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicIntegerFieldUpdater;
+import java.util.concurrent.atomic.AtomicLongFieldUpdater;
+import java.util.concurrent.atomic.AtomicReferenceFieldUpdater;
+import java.util.function.BiFunction;
+
+import static com.azure.core.amqp.implementation.ClientConstants.DELIVERY_STATE_KEY;
+import static com.azure.core.amqp.implementation.ClientConstants.ENTITY_PATH_KEY;
+import static com.azure.core.amqp.implementation.ClientConstants.LINK_NAME_KEY;
+import static com.azure.core.amqp.implementation.ClientConstants.DELIVERY_TAG_KEY;
+import static com.azure.core.util.FluxUtil.monoError;
+
+/**
+ * The Flux operator to stream messages reliably from a messaging entity (e.g., Event Hub partition,
+ * Service Bus Queue, Topic) to downstream message subscriber.
+ */
+public final class MessageFlux extends FluxOperator<AmqpReceiveLink, Message> {
+    private static final String MESSAGE_FLUX_KEY = "messageFlux";
+    private final ClientLogger logger;
+    private final int prefetch;
+    private final CreditFlowMode creditFlowMode;
+    private final AmqpRetryPolicy retryPolicy;
+    private volatile BiFunction<String, DeliveryState, Mono<Void>> updateDispositionFunc;
+
+    /**
+     * Create a message-flux to stream messages from a messaging entity to downstream subscriber.
+     *
+     * @param source the upstream source that, upon a request, provide a new receiver connected to the messaging entity.
+     * @param prefetch the number of messages that the operator should prefetch from the messaging entity (for a
+     *                 less chatty network and faster message processing on the client).
+     * @param creditFlowMode the mode indicating how to compute the credit and when to send it to the broker.
+     * @param retryPolicy the retry policy to use to obtain a new receiver upon current receiver termination.
+     */
+    public MessageFlux(Flux<? extends AmqpReceiveLink> source, int prefetch, CreditFlowMode creditFlowMode, AmqpRetryPolicy retryPolicy) {
+        super(source);
+
+        final Map<String, Object> loggingContext = new HashMap<>(1);
+        loggingContext.put(MESSAGE_FLUX_KEY, StringUtil.getRandomString("mf"));
+        this.logger = new ClientLogger(MessageFlux.class, loggingContext);
+
+        if (prefetch < 0) {
```

Comment:
Might be worthwhile to add these throws to the Javadocs since I see you added params etc.

Reply:
Yes, good idea, I've updated the java-doc.